cmake: export targets, allow to build models via CMake

matz-e opened the pull request:
This PR aims to allow dependent projects to build their models via CMake
rather than a combination of shell scripts and Makefiles.  In the long
term, this may help facilitate building natively on Windows.

To try:
```
git clone -b imported-main https://github.com/BlueBrain/neurodamus-models.git
cd neurodamus-models
```

Create a `CMakeLists.txt` with contents like:
```cmake
cmake_minimum_required(VERSION 3.28)

project(newrodamus)

find_package(neuron REQUIRED)

create_nrnmech(
  CORENEURON  # Will enable a coreneuron build
  MOD_FILES
  neocortex/mod/v6/CaDynamics_DC0.mod
  neocortex/mod/v6/Ca_HVA2.mod
  neocortex/mod/v6/Ca_LVAst.mod
  neocortex/mod/v6/DetAMPANMDA.mod
  neocortex/mod/v6/DetGABAAB.mod
  neocortex/mod/v6/GluSynapse.mod
  neocortex/mod/v6/Ih.mod
  neocortex/mod/v6/K_Pst.mod
  neocortex/mod/v6/K_Tst.mod
  neocortex/mod/v6/KdShu2007.mod
  neocortex/mod/v6/NaTg.mod
  neocortex/mod/v6/Nap_Et2.mod
  neocortex/mod/v6/ProbAMPANMDA_EMS.mod
  neocortex/mod/v6/ProbGABAAB_EMS.mod
  neocortex/mod/v6/SK_E2.mod
  neocortex/mod/v6/SKv3_1.mod
  neocortex/mod/v6/StochKv3.mod
  neocortex/mod/v6/TTXDynamicsSwitch.mod
  neocortex/mod/v6/VecStim.mod
  neocortex/mod/v6/gap.mod
  neocortex/mod/v6/netstim_inhpoisson.mod
)
```

Then build and install:
```
cmake -B build -S . -GNinja -DCMAKE_INSTALL_PREFIX=x86_64
cmake --build build
cmake --install build
```

See BlueBrain/neurodamus-models#12 for an example use case.

To be continued‥

CI_BRANCHES:NMODL_BRANCH=cmake-exports

matz-e committed:
cmake: export targets, allow to build models via CMake

This PR aims to allow dependent projects to build their models via CMake
rather than a combination of shell scripts and Makefiles.  In the long
term, this may help facilitate building natively on Windows.

To try:
```
git clone -b imported-main https://github.com/BlueBrain/neurodamus-models.git
cd neurodamus-models
```

Create a `CMakeLists.txt` with contents like:
```cmake
cmake_minimum_required(VERSION 3.28)

project(newrodamus)

find_package(neuron REQUIRED)

create_libnrnmech(MOD_FILES
  neocortex/mod/v6/CaDynamics_DC0.mod
  neocortex/mod/v6/Ca_HVA2.mod
  neocortex/mod/v6/Ca_LVAst.mod
  neocortex/mod/v6/DetAMPANMDA.mod
  neocortex/mod/v6/DetGABAAB.mod
  neocortex/mod/v6/GluSynapse.mod
  neocortex/mod/v6/Ih.mod
  neocortex/mod/v6/K_Pst.mod
  neocortex/mod/v6/K_Tst.mod
  neocortex/mod/v6/KdShu2007.mod
  neocortex/mod/v6/NaTg.mod
  neocortex/mod/v6/Nap_Et2.mod
  neocortex/mod/v6/ProbAMPANMDA_EMS.mod
  neocortex/mod/v6/ProbGABAAB_EMS.mod
  neocortex/mod/v6/SK_E2.mod
  neocortex/mod/v6/SKv3_1.mod
  neocortex/mod/v6/StochKv3.mod
  neocortex/mod/v6/TTXDynamicsSwitch.mod
  neocortex/mod/v6/VecStim.mod
  neocortex/mod/v6/gap.mod
  neocortex/mod/v6/netstim_inhpoisson.mod
)
```

Then build and install:
```
cmake -B build -S . -GNinja -DCMAKE_INSTALL_PREFIX=x86_64
cmake --build build
cmake --install build
```

To be continued‥

matz-e committed:
Bump stuff.

github-actions[bot] committed:
Fix formatting

Commit:
Merge branch 'master' into export-cmake

Commit:
Add fmt to linked libs for testing

Commit:
Also link fmt in Python build

JCGoran commented:
I merged `master` (after resolving conflicts) and added some minimal improvements to make it build successfully.
I think the remainder of the failures requires setting the right optimization flags since the test reports:

```plaintext
DIFF at line 1: -66.0883!=-66.0883
```

Commit:
Put back the original `hh.mod`

Commit:
Fix issue with corenrn target

Commit:
Merge branch 'master' into export-cmake

Commit:
Fix issues with paths

Always use resolved absolute paths in CMake since nocmodl has some
trouble with symlinks.

Commit:
Experiment with replacing `nrnivmodl`

The `nrnivmodl` script is...complicated. With the move to building mod
files using pure CMake, it makes sense to simplify it to the maximum
extent possible. This commit makes the following changes to how
`nrnivmodl` works now:

- there is a cmake file, `nrnivmodl.cmake`, which builds all of the mod
  files. The list of mod files is passed from the bash script as a cmake
definition, so that everything is delegated to cmake as soon as possible
- the above file is currently `configure_file`-able; maybe an API
  version would be useful to add?
- the `nrnivmodl.cmake` file is installed alongside NEURON
- the `nrnivmodl.in` script is `configure_file`-able; only one parameter
  has practical significance, and that is the @CORENEURON@ one (set to
an empty string if using in `nrnivmodl` mode, and to "CORENEURON" if
using in `nrnivmodl-core` mode). It first tries to copy the `nrnivmodl.cmake` file to
the current directory; if the current directory already has a
CMakeLists.txt that differs from that one, it errors out. If the two are
identical, it runs the equivalent of `cmake -B [ARCH] -DMODFILES="$@"`.
Depending on whether or not the script is called `nrnivmodl` or
`nrnivmodl-core`, it also adds the `-DCORENEURON=CORENEURON` argument.
The script also sets `CMAKE_PREFIX_PATH` to where NEURON is
installed (so it can find NEURON using the `neuronConfig.cmake` file),
but has the option of overriding this behavior (for purposes of
debugging).

TODO list:
- I still need to remove the original `nrnivmodl-core` because it is
  rather complicated (do we need so many options?)
- remove `nrnivmodl_makefile` and co. so we are finally free of
  hand-made makefiles

Commit:
Add mod files as dependencies

TODO figure out what to do with the NMODL `INCLUDE` statement (see:
https://nrn.readthedocs.io/en/latest/nmodl/language/nmodl.html#include).

Commit:
Add `TARGET_NAME` and `OUTPUT_DIR` to create_nrnmech

Commit:
Add edge case of a directory

Commit:
Some refactoring of `create_nrnmech`

Commit:
Fix issue with modfile targets

Turns out, `.inc` files with the same name as a `.mod` file was causing
problems

Commit:
Fix formatting and add better nocmodl detection

Commit:
Add warning message for nrnivmodl when files have different ext

Commit:
Put nrnmech in `.libs` subdir

Commit:
Add neurondemo `special` binary

Commit:
Fix nrnivmodl script again

Commit:
WIP on getting coreneuron stuff to work

Commit:
Replace OUTPUT_DIR with ARTIFACTS, LIBRARY, and EXECUTABLE

Gives more flexibility to users

Commit:
Fix issue with generation of mod_reg file

Commit:
Minor tweaks to get the tests to work

Commit:
Add variable helper and remove makefile cruft

Commit:
Remove `run_test.sh`

Commit:
Formatting

Commit:
Do not call neurondemo since we install it normally

Commit:
Changes so coreneuron stuff works too

Commit:
Only enable certain tests if coreNEURON is enabled

Commit:
Forgot to include dir

Commit:
Fix issue with path of demo

Commit:
tqperf also needs coreneuron it seems

Commit:
Fix nmodl tests

Quoting issues again...

Commit:
[REMOVE] Upload build artifacts

Commit:
Slashes...

Commit:
[REMOVE] Cache pip packages

Commit:
[REMOVE] add build file list

Commit:
[REMOVE] YAML

Commit:
[REMOVE] Cache paths

Commit:
Include project binary dir when building demo

Commit:
Just add nrnivmodl_dependency for the demo

Commit:
Possible fix for tqperf test

Commit:
Make builtin-libcorenrnmech an alias for the mod files

Commit:
Fix tqperf tests

Commit:
Remove `NRNIVMODL_ARGS` in favor of `EXTRA_INCLUDES` and `EXTRA_LIBRARIES`

Commit:
Formatting again

Commit:
Fix all useless stuff so we can run tests from build dir

`nrnivmodl` is not very configure-able now, so we make-do with copying
a bunch of useless files to the build directory.

Commit:
Better nrnivmodl

Commit:
Fix install issue

Commit:
Merge branch 'master' into export-cmake

Commit:
Fix some build issues

Commit:
Set mechanism suffix to module

NMODL mechanisms are actually modules, but to keep things simple, we
build them as shared libraries. This also means that the suffix must be
adjusted on MacOS, since a shared library has a non-.so suffix.

Commit:
Fix NMODL usecase tests

Commit:
Quoting and cmake

Commit:
Fix install location for some targets

Commit:
Fix undefined variables in CI script

Commit:
Fix build issues from pyproject migration

Commit:
Merge branch 'jelic/fix_build_issues' into export-cmake

Commit:
Slight fixes to install dirs

Commit:
Merge branch 'master' into export-cmake

Commit:
Update docs for create_nrnmech

Commit:
Fix macos deployment target issue

Commit:
Fix suffix and rpath shenanigans

Commit:
Fixes to cmake paths and removed old `nrnivmodl-core`

Commit:
Fix most of the remaining issues

Commit:
Better handling of `nrnivmodl[-core,-all]`

Commit:
Fix install issue for nrnivmodl-core

Commit:
Always use CMAKE_SHARED_LIBRARY_SUFFIX

Commit:
Fix auditwheel fixing

Also fix test for MacOS

Commit:
Use Ninja if it's available and CMAKE_GENERATOR is not overridden

Commit:
Merge branch 'master' into export-cmake

Commit:
Install lib and scripts

Commit:
Disable ASAN when building mechanisms

NMODL and NOCMODL are too noisy for now

Commit:
Fix issue with nrniv-core

Also fix NMODL install paths of binaries

Commit:
Fix issue with nrnivmodl when running under Python

The problem is that when running nrnivmodl under the Python wrapper, it
creates a `special`, but then `special` is actually a Python script
which wraps the `special.nrn` executable.

Commit:
Rerun CI

Commit:
Merge branch 'master' into export-cmake

Commit:
Fix issue with install paths

Should more or less work now.

Commit:
Fix slashes again

Commit:
Remove unrelated CI part

Commit:
Un-update CLI11

Commit:
Fixes so it compiles on NVHPC

TODO figure out linkage to libcudart

Commit:
Fix docs dependency

We need the NMODL binary to build the notebooks because one of them
calls nrnivmodl, which needs to find NMODL

Commit:
Use CMAKE_INSTALL_PREFIX on Windows

Commit:
Apply symlink fix

See #3405

Commit:
Merge branch 'master' into export-cmake

Commit:
Merge branch 'master' into export-cmake

Commit:
Add mod directories themselves as include ones

Commit:
Merge branch 'master' into export-cmake

Commit:
Fix install dir for NMODL lexer and parser

Commit:
Merge remote-tracking branch 'origin/export-cmake' into export-cmake

Commit:
Merge branch 'master' into export-cmake
